compiler/noirc_evaluator/src/ssa/opt/array_set.rs

```diff
@@ -133,10 +139,12 @@ impl<'f> Context<'f> {
                         }
                     }
                 }
-                Instruction::Load { .. } => {
+                Instruction::Load { address } => {
                     let result = self.dfg.instruction_results(*instruction_id)[0];
                     if matches!(self.dfg.type_of_value(result), Array { .. } | Slice { .. }) {
-                        self.arrays_from_load.insert(result);
+                        let is_reference_param =
+                            self.dfg.block_parameters(block_id).contains(address);
+                        self.arrays_from_load.insert(result, is_reference_param);
```

Comment:
Can this ever be transitive? 

I'm looking at this initial SSA of a push:
```
acir(inline) fn push f2 {
  b0(v0: &mut [Field; 3], v1: &mut u32, v2: Field):
    v3 = load v0
    inc_rc v3
    v4 = load v0
    v5 = load v1
    v7 = lt v5, u32 3
    constrain v7 == u1 1 '"push out of bounds"'
    v9 = load v0
    inc_rc v9
    v10 = load v0
    v11 = load v1
    v12 = load v0
    v13 = load v1
    v14 = array_set v10, index v13, value v2
    v16 = add v13, u32 1
    store v14 at v0
    store v11 at v1
    v17 = load v0
    inc_rc v17
    v18 = load v0
    v19 = load v1
    v20 = load v0
    v21 = load v1
    v22 = add v21, u32 1
    store v18 at v0
    store v22 at v1
    v23 = load v0
    dec_rc v23
    return 
}
```
I can see that `v0` is loaded multiple times, for example:
```
    v10 = load v0
    v12 = load v0
    v13 = load v1
    v14 = array_set v10, index v13, value v2
    store v14 at v0
    v17 = load v0
    inc_rc v17
    v18 = load v0
```
I'm wondering if there was another `array_set` on, say, `v17` on `v18`, would that be mutable because it's not an input to `b0` like `v0` is?

Reply:
This checks whether the load of the address comes from a parameter. So `v0` would still be what is checked for `v17` and `v18`.